### PR TITLE
Add initial pixi.toml

### PR DESCRIPTION
**Pull request by certik**
Run with:

    pixi run build
    pixi run test

### CODE REVIEW

**Comment by certik** [comment]
The build runs about 3x slower than without pixi. I am trying to understand why. First I applied the following patch:
```diff
diff --git a/pixi.toml b/pixi.toml
index 9c4fdd34f..4529d6d9a 100644
--- a/pixi.toml
+++ b/pixi.toml
@@ -12,14 +12,13 @@ build0 = "bash ./build0.sh"
 [tasks.build]
 cmd = """
     cmake . \
-        -GNinja \
         -DCMAKE_BUILD_TYPE=Debug \
         -DLFORTRAN_BUILD_ALL=yes \
         -DWITH_LLVM=yes \
         -DWITH_STACKTRACE=yes \
         -DCMAKE_PREFIX_PATH=$CONDA_PREFIX \
         -DCMAKE_INSTALL_PREFIX=$PIXI_PACKAGE_ROOT/inst &&
-    cmake --build . --parallel --target install
+    cmake --build . --parallel 8 --target install
     """
 
 [tasks.test]
@@ -45,7 +44,7 @@ xonsh = "*"
 re2c = "*"
 numpy = "*"
 rapidjson = "*"
-ninja = "*"
+make = "*"
 zlib = "*"
 
 [target.osx-arm64.dependencies]
```
This uses `make` that I usually use, but it still takes 1.5 minutes:
```
pixi run build  399.26s user 12.11s system 486% cpu 1:24.47 total
```
Compared to the native build:
```
./build1.sh  131.97s user 6.45s system 430% cpu 32.173 total
```

The main difference that I can see is the compiler. Pixi uses:
```
C++ compiler    : /Users/ondrej/repos/lfortran/lfortran/.pixi/env/bin/arm64-apple-darwin20.0.0-clang++
```
While I use:
```
C++ compiler    : /Library/Developer/CommandLineTools/usr/bin/c++
```

So maybe the pixi compiler is just slow. @wolfv do you know if the Pixi compiler is compiled in Release mode, to compile quickly?

**Comment by certik** [comment]
I tried this patch:
```diff
--- a/pixi.toml
+++ b/pixi.toml
@@ -11,15 +11,14 @@ build0 = "bash ./build0.sh"
 
 [tasks.build]
 cmd = """
-    cmake . \
-        -GNinja \
+    CC=/usr/bin/cc CXX=/usr/bin/c++ cmake . \
         -DCMAKE_BUILD_TYPE=Debug \
         -DLFORTRAN_BUILD_ALL=yes \
         -DWITH_LLVM=yes \
         -DWITH_STACKTRACE=yes \
         -DCMAKE_PREFIX_PATH=$CONDA_PREFIX \
         -DCMAKE_INSTALL_PREFIX=$PIXI_PACKAGE_ROOT/inst &&
-    cmake --build . --parallel --target install
+    cmake --build . --parallel 8 --target install
     """
 
 [tasks.test]
@@ -45,7 +44,7 @@ xonsh = "*"
 re2c = "*"
 numpy = "*"
 rapidjson = "*"
-ninja = "*"
+make = "*"
 zlib = "*"
 
 [target.osx-arm64.dependencies]
```
But still slow:
```
pixi run build  382.15s user 8.98s system 475% cpu 1:22.20 total
```
So it's not the compiler. It must be some libraries or compiler options then. Let me investigate.

**Comment by certik** [comment]
Running `make VERBOSE=1` prints for Pixi:
```
[  2%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o
cd /Users/ondrej/repos/lfortran/lfortran/src/libasr && /Library/Developer/CommandLineTools/usr/bin/c++ -DHAVE_TARGET_AARCH64=1 -DHAVE_WHEREAMI=1 -DHAVE_ZLIB=1 -DLCOMPILERS_FAST_ALLOC=1 -I/Users/ondrej/repos/lfortran/lfortran/src/libasr/.. -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/ondrej/repos/lfortran/lfortran/.pixi/env/include -Wall -Wextra -g -ggdb -std=gnu++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -fPIC -fno-rtti -MD -MT src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o -MF CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o.d -o CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o -c /Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_cpp.cpp
```
And my build:
```
[  2%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o
cd /Users/ondrej/repos/lfortran/lfortran/src/libasr && /Library/Developer/CommandLineTools/usr/bin/c++ -DHAVE_TARGET_AARCH64=1 -DHAVE_WHEREAMI=1 -DHAVE_ZLIB=1 -DLCOMPILERS_FAST_ALLOC=1 -I/Users/ondrej/repos/lfortran/lfortran/src/libasr/.. -isystem /Users/ondrej/mambaforge/envs/lf/include -Wall -Wextra -g -ggdb -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk -fPIC -fno-rtti -std=gnu++17 -MD -MT src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o -MF CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o.d -o CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o -c /Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_cpp.cpp
```

Comparing these two I can see that Pixi added the following options: `-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++`. So that explains it, for some reason it gets compiled in "semi-release" mode, so it's slow.

**Comment by certik** [comment]
Using `pixi shell` I can see that it is supplying:
```
CXXFLAGS=-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/ondrej/repos/lfortran/lfortran/.pixi/env/include
```
@wolfv I can disable this in `pixi.toml`, but I assume you need some of the other options. Is there a way to make Pixi understand Release / Debug modes?

When developing I want to build without optimizations so that it builds quickly.

File: pixi.toml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+xtl = "*"
+nlohmann_json = "*"
+cppzmq = "*"
+jupyter_kernel_test = "*"
+xonsh = "*"
+re2c = "*"
+numpy = "*"
+rapidjson = "*"
+ninja = "*"
+zlib = "*"
```
Comment (HaoZeke) [inline pixi.toml:49]: Isn't it better to have these pin resolved dependencies on their own via `pixi add`?

**Comment by wolfv** [comment]
Right, these come from the "compiler" activation scripts that conda-forge uses to maximize compatibility with all libraries. I'll think of some solutions. One would be to override CPPFLAGS of course

**Comment by certik** [comment]
@wolfv excellent, thanks. For now I fixed it in 974f03b2dd0528ee411c31814195a7d5dbdf27c6. Now it is much faster:
```
pixi run build  206.06s user 69.95s system 649% cpu 42.497 total
```
It's still 10s slower, but it's using all the defaults from pixi (the pixi C++ compiler, etc.), so it's probably some other options or so.

I think this is good enough for an initial PR.

In general, when developing I want fast compilation and enable asserts in LFortran. But when building a conda package, I want to build in Release mode which disables asserts. Pixi should probably support this Debug / Release workflow somehow.

**Comment by wolfv** [comment]
Yes, one more reason why pixi is slower than "native" is that there is an "environment activation" step where certain scripts are executed and environment variables are set up. This is what would usually happen with `conda activate`. 

In pixi we already execute the activation seperately, extract the environment variables and then set them for whatever shell is used. We are not caching the result of this activation though, which could be a potential speed-improvement for the future.

**Comment by certik** [comment]
I am going to go ahead and merge this. We can iteratively improve this as we go. Thanks @wolfv !

The CI failure is unrelated (https://github.com/lfortran/lfortran/issues/2442).

**Pull request merged**